File: backend/src/utils/billing.ts
import moment from 'moment';

export type BillingFrequency = 'daily' | 'weekly' | 'monthly' | 'annually';

export const calculateNextBillingDate = (
  currentDate: Date | string,
  frequency: BillingFrequency,
  billingDay: number
): Date => {
  const now = moment(currentDate).startOf('day');
  let nextDate = moment(now);

  if (frequency === 'daily') {
    nextDate.add(1, 'days');
  } else if (frequency === 'weekly') {
    nextDate.isoWeekday(billingDay % 7 || 7);
    if (nextDate.isSameOrBefore(moment(now))) {
      nextDate.add(1, 'weeks').isoWeekday(billingDay % 7 || 7);
    }
  } else if (frequency === 'monthly') {
    nextDate.date(billingDay);
    if (nextDate.isSameOrBefore(moment(now))) {
      nextDate.add(1, 'months').date(billingDay);
    }
  } else if (frequency === 'annually') {
    nextDate.add(1, 'years').date(billingDay);
    if (nextDate.isSameOrBefore(moment(now))) {
      nextDate.add(1, 'years').date(billingDay);
    }
  }

  return nextDate.startOf('day').toDate();
};

File: backend/app/utils/billing.py
from datetime import datetime, timedelta


def calculate_next_billing_date(
    current_date: datetime, frequency: str, billing_day: int
) -> datetime:
    if frequency == "daily":
        return current_date + timedelta(days=1)
    elif frequency == "weekly":
        return current_date + timedelta(weeks=1)
    elif frequency == "monthly":
        return _add_months(current_date, 1, billing_day)
    elif frequency == "quarterly":
        return _add_months(current_date, 3, billing_day)
    elif frequency == "annually":
        return _add_months(current_date, 12, billing_day)
    return current_date + timedelta(days=30)


def _add_months(date: datetime, months: int, day_of_month: int) -> datetime:
    month = date.month - 1 + months
    year = date.year + month // 12
    month = month % 12 + 1
    day = min(day_of_month, _days_in_month(year, month))
    return date.replace(year=year, month=month, day=day)


def _days_in_month(year: int, month: int) -> int:
    if month == 12:
        return 31
    next_month = datetime(year, month + 1, 1)
    last_day = next_month - timedelta(days=1)
    return last_day.day
